src/rust/engine/process_execution/src/remote_cache.rs

```diff
@@ -117,18 +119,13 @@ impl CommandRunner {
         Component::Normal(name) => name
           .to_str()
           .ok_or_else(|| format!("unable to convert '{:?}' to string", name))?,
-        _ => return Err("illegal state: unexpected path component in relative path".into()),
+        _ => return Ok(None),
       };
 
       // Load the Directory proto corresponding to `current_directory_digest`.
       let current_directory = match store.load_directory(current_directory_digest).await? {
         Some((dir, _)) => dir,
-        None => {
-          return Err(format!(
-            "illegal state: directory for digest {:?} did not exist locally",
-            &current_directory_digest
-          ))
-        }
+        None => return Ok(None),
```

Comment:
I'm not following the code well, but should this be mixed-mode? IOW if the store doesn't have the digest is that an Err and only the case where the current directory doesn't have the path component the new not Err but None state?

Reply:
Yea, that sounds right. This shouldn't hide failures to load `Digests` if they have been declared.

But the error messages for these aren't quite accurate: `load_directory` will backfill from the remote as well, so "did not exist locally" could just be "did not exist in the store".

Reply:
Errors are already propagated out of the function. Note the use of the `?` operator after the call to `load_directory`. The `match` is only dealing with existence of the digest.

Reply:
`load_directory` will return `None` if the directory does not exist: but a missing directory here would be a `Digest` which was not stored. That should be an error in this case.

Reply:
Will be fixed by https://github.com/pantsbuild/pants/pull/11775.